### ISSUE TITLE
Parallel Binary search for NEURON.

### ISSUE THREAD

**Issue by russelljjarvis** [scrook/neuroml-db#112]
Hi Justas,

Where is the other binary search we were talking about in this code?

https://github.com/scrook/neuroml-db/blob/master/Import%20Scripts/model-importer/cellmodel.py#L753

```
def list_process(iterable_arg):
    iterable_arg +=1
    return iterable_arg
```

```
cpus = multiprocessing.cpu_count -1
scattered_iterable = db.from_sequence(list,npartitions=cpus)

```
Results of parallel map are implicitly gathered back onto CPU0, the main thread.
```
gathered_list = scattered_iterable.map(list_process)
```
Russell.

**Comment by russelljjarvis** [comment]
```    
    def recompute_range(iteration,upperLevel,lowerLevel):
        intervals = np.arange(upperLevel,lowerLevel,cpus)
        return intervals

    def list_process(simulate_iteration):
        runner = NeuronRunner(simulate_iteration)
        try:
            found = runner.run()
            return found
        except NumericalInstabilityException:
            if on_unstable is not None:
                found = on_unstable()
                return found
            else:
                return None

    def simulate_iteration(time_flag):
        self.time_flag = time_flag
        h = self.build_model()
        self.restore_state(state_file=state_file)

        self.setCurrent(amp=currentAmp, delay=current_delay, dur=current_duration)
        print("Trying " + str(currentAmp) + " nA...")

        if not test_early:
            t, v = self.runFor(run_for_after_delay)
            found = test_condition(t, v)
        else:
            t, v = self.runFor(run_for_after_delay, test_condition)
            found = test_condition(t, v)

        plt.plot(t, v, label=str(round(currentAmp, 4)) + ", Found: " + str(found))
        plt.legend(loc='upper left')
        plt.savefig(str(iteration) + " " + fig_file)

        print("FOUND" if found else "NOT FOUND")

        return found



    def find_border(self, lowerLevel, upperLevel,
                    current_delay, current_duration,
                    run_for_after_delay, test_condition, max_iterations, fig_file,
                    skip_current_delay=False, on_unstable=None, test_early=False):

        state_file = 'border_state.bin'

        if not skip_current_delay:
            def reach_resting_state(time_flag):
                self.time_flag = time_flag
                self.build_model()

                print("Simulating till current onset...")
                self.sim_init()
                self.setCurrent(amp=0, delay=current_delay, dur=current_duration)
                self.runFor(current_delay)
                self.save_state(state_file=state_file)
                print("Resting state reached. State saved.")

            runner = NeuronRunner(reach_resting_state)
            result = runner.run()

        iterate = True
        iteration = 0
        found_once = False

        upperLevel_start = upperLevel
        lowerLevel_start = lowerLevel

        first_interval = np.arange(lowerLevel_start,upperLevel_start,cpus)
        scattered_iterable = db.from_sequence(first_interval,npartitions=cpus)
        gathered_list = scattered_iterable.map(list_process)
        while iterate:
            scattered_iterable = db.from_sequence(intervals,npartitions=cpus)
            gathered_list = scattered_iterable.map(list_process)
            for g in gathered_list:
                if g == True:
                    break
            intervals = recompute_range(iteration,upperLevel,lowerLevel)
            if found:
                upperLevel = currentAmp
                found_once = True
            else:
                lowerLevel = currentAmp

            iteration = iteration + 1

            if iteration >= max_iterations or lowerLevel == upperLevel_start or upperLevel == lowerLevel_start:
                iterate = False
        current_range = (lowerLevel, upperLevel)
        return current_range, found_once
```

**Comment by russelljjarvis** [comment]
> ```python
>     import multiprocessing
>     cpus = multiprocessing.cpu_count
>     import dask.bag as db # a pip installable module, usually installs without complication
>     import dask    
>     def recompute_range(iteration,upperLevel,lowerLevel):
>         intervals = np.arange(upperLevel,lowerLevel,cpus)
>         return intervals
> 
>     def list_process(simulate_iteration):
>         runner = NeuronRunner(simulate_iteration)
>         try:
>             found = runner.run()
>             return found
>         except NumericalInstabilityException:
>             if on_unstable is not None:
>                 found = on_unstable()
>                 return found
>             else:
>                 return None
> 
>     def simulate_iteration(time_flag):
>         self.time_flag = time_flag
>         h = self.build_model()
>         self.restore_state(state_file=state_file)
> 
>         self.setCurrent(amp=currentAmp, delay=current_delay, dur=current_duration)
>         print("Trying " + str(currentAmp) + " nA...")
> 
>         if not test_early:
>             t, v = self.runFor(run_for_after_delay)
>             found = test_condition(t, v)
>         else:
>             t, v = self.runFor(run_for_after_delay, test_condition)
>             found = test_condition(t, v)
> 
>         plt.plot(t, v, label=str(round(currentAmp, 4)) + ", Found: " + str(found))
>         plt.legend(loc='upper left')
>         plt.savefig(str(iteration) + " " + fig_file)
> 
>         print("FOUND" if found else "NOT FOUND")
> 
>         return found
> 
> 
> 
>     def find_border(self, lowerLevel, upperLevel,
>                     current_delay, current_duration,
>                     run_for_after_delay, test_condition, max_iterations, fig_file,
>                     skip_current_delay=False, on_unstable=None, test_early=False):
> 
>         state_file = 'border_state.bin'
> 
>         if not skip_current_delay:
>             def reach_resting_state(time_flag):
>                 self.time_flag = time_flag
>                 self.build_model()
> 
>                 print("Simulating till current onset...")
>                 self.sim_init()
>                 self.setCurrent(amp=0, delay=current_delay, dur=current_duration)
>                 self.runFor(current_delay)
>                 self.save_state(state_file=state_file)
>                 print("Resting state reached. State saved.")
> 
>             runner = NeuronRunner(reach_resting_state)
>             result = runner.run()
> 
>         iterate = True
>         iteration = 0
>         found_once = False
> 
>         upperLevel_start = upperLevel
>         lowerLevel_start = lowerLevel
> 
>         first_interval = np.arange(lowerLevel_start,upperLevel_start,cpus)
>         scattered_iterable = db.from_sequence(first_interval,npartitions=cpus)
>         gathered_list = scattered_iterable.map(list_process)
>         while iterate:
>             scattered_iterable = db.from_sequence(intervals,npartitions=cpus)
>             gathered_list = scattered_iterable.map(list_process)
>             for g in gathered_list:
>                 if g == True:
>                     break
>             intervals = recompute_range(iteration,upperLevel,lowerLevel)
>             if found:
>                 upperLevel = currentAmp
>                 found_once = True
>             else:
>                 lowerLevel = currentAmp
> 
>             iteration = iteration + 1
> 
>             if iteration >= max_iterations or lowerLevel == upperLevel_start or upperLevel == lowerLevel_start:
>                 iterate = False
>         current_range = (lowerLevel, upperLevel)
>         return current_range, found_once
> ```
Yes, okay, I think that code would only be marginally different. Would it be hard for me to build the stack so I can test the code?

**Comment by JustasB** [comment]
You should be able to clone the https://github.com/scrook/neuroml-db repo. But it's rather large 13GB+ and most of the relevant code is just in this folder: https://github.com/scrook/neuroml-db/tree/master/Import%20Scripts/model-importer

I have never used the code to do things in isolation (only with respect to models in NML-DB), but you might be able to get what you need from manager and manage classes.

[check_install_dependencies method](https://github.com/scrook/neuroml-db/blob/master/Import%20Scripts/model-importer/manage.py#L109) checks/installs dependencies needed for running NML-DB updates, but you should be able to pick and chose and install dependencies needed for your part of the code.

[save_model_properties method ](https://github.com/scrook/neuroml-db/blob/master/Import%20Scripts/model-importer/manager.py#L933) is the primary way how I interface with the DB, get the models, and compute their properties
